accepted/00062-hub-xmlrpc-api.md

```diff
@@ -0,0 +1,129 @@
+      * `hub.login(username, password)` → `hubSessionKey`
+      * `hub.attachToServer(hubSessionKey, serverId, username, password)` → `serverSessionKey`
+      * `hub.serverMethod(serverSessionKey, parameters)` → `output`
+      * N+1 pairs of credentials will be needed (one for the Hub, one per each of the N Servers)
+     * Authentication relay mode: as an additional convenience option, which can be activated with a flag, Hub-User credentials can be re-used to authenticate against Servers. So if a certain username/password is uniformly configured on the Hub and across several Servers, Gateway usage becomes simpler: only one log in is needed, with its Hub-User credentials
+     * Automatic connect mode: as an additional convenience option, which can be activated with a flag, the Hub gateway API can automatically select the list of Servers a Hub-User works with, based on the list of Servers that same Hub-User has permissions on
+       - note: that this only makes sense if users are OK with the idea of using Server sysadmin users to act on Clients
+       - If they prefer to manage Clients via users that have no permissions to manage Servers, they can create a Hub-Org with no Servers. Any Hub-User in this Hub-Org will be able to connect to the Gateway, and possibly have permissions on Clients (if Servers have Server-Users with same name and passwords)
+
+* Multicasting: call an XMLRPC method on _N_ Servers at once
+  * method name stays the same
+  * Assuming the method on the Server accepts _M_ parameters, the same method on the Hub will accept an array of _N_ parameters, each of which contains as elements the _M_ parameters for a given Server
+    * eg. `actionchain.listChains(sessionKey)` → `hub.actionchain.listChains([serverSessionKey1, serverSessionKey2, ...])`
+    * eg. `channel.org.enableAccess(sessionKey, channelLabel, orgId)` → `hub.channel.org.enableAccess([serverSessionKey1, channelLabel1, orgId1], [serverSessionKey2, channelLabel2, orgId2], ...)`
+  * return types would be arrays of _N_ elements
+    * special elements would be needed to signal Exceptions
+
+* Availability
+  * if a Server is down at the time a Gateway call targeting it is made, call should fail after a configurable timeout. If multiple Servers are targeted, only that call fails and others continue
+
+
+## Impact on existing components and users
+
+All code would be new in a new component, so no change to existing components is expected. No impact on existing users is expected at all.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+We are not adding any Hub-centered functionality basically, this is a little more than a router.
+
+## Limitations
+
+- implementation of API methods currently not implemented on the Server, eg. downstream [#2361](https://github.com/SUSE/spacewalk/issues/2361) [#2362](https://github.com/SUSE/spacewalk/issues/2362), is explicitly not covered by this RFC
+
```

Comment:
You may need some minimal error handling code in case the API call that was multicast isn't available on all nodes.

Reply:
Right, I did not state the assumption here is that all Servers are on the same version of Manager, at least major, where we already guarantee no breaking changes. Added now.